cpp/include/doxygen_groups.h

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file
+ * @brief Doxygen group definitions
+ */
+
+// This header is only processed by doxygen and does
+// not need to be included in any source file.
+// Below are the main groups that doxygen uses to build
+// the Modules page in the specified order.
+//
+// To add a new API to an existing group, just use the
+// @ingroup tag to the API's doxygen comment.
+// Add a new group by first specifying in the hierarchy below.
```

Comment:
Where can we add more documentation for each group? These descriptions are very brief.

Reply:
Great question, I believe doxygen is very liberal with where the group documentation should be. You can write the docs anywhere inside the group, which includes the lines immediately after the `@defgroup` or anywhere within the group block (between `@{` and `@}`).

Libcudf adds module description by in-lining them after each `@addtogroup` decorator.
Unlike libcudf where the modules are quite well organized with the file structures, libcuspatial is a bit fragmented atm. For example the distances computation API are scattered in several header files. With this in mind, I moved all group descriptions to `doxygen_groups.h`. This provides a nice overview to the structure of the entire code base and reduces large group docs inlined within codes. Let me know if the new organization works better.